wip; Graph Adaptation for the Java Driver

newkek opened the pull request:
This is a wip but I'm opening a PR so we can directly centralize comments and remarks here.

The current _graph_ branch, at the time this PR is open, is based on the Java Driver 3.0.0-alpha4 version, and needs to be rebased on the latest 3.x version available.

Commit:
Improvemments on the results API, corrected setGraphXXX() methods

Commit:
Switched JSON parsing to Jackson library

- added 'asXXX()' methods on GraphData where XXX is a type handled in JSON (int, bool, String)
- optimized imports

Commit:
Added support for GraphStatements with values

Notes : - see if the GraphStatements constructor with values is going to be used
- Changed "checkStatement" static method's class

Commit:
Small bug fixes to Statements and Statements values.

Commit:
wip; refactor of Statements because of GraphStatements inheritance

Commit:
wip; Refactoring graph statements model, simple GraphStatements working.

Commit:
Refactored for 3.0 API changes...

- Refactored Bound and Prepared Graph statements.
- Added options to set default values to Graph payload values through the session
	|-> new methods to directly execute and prepare a query string through the session

Commit:
First graph prototype

Commit:
Added GraphTraversalResult.asVertex/asEdge()

- You can now get a GraphTraversalResult directly transformed as a Vertex or Edge object.
- You can also plug your very own cherished Vertex or Edge implementation by using asVertex/asEdge(CustomVertexClass.class), providing also your implementation of the GraphJsonDeserializer<CustomVertexClass> or extending the default ones and adding your fashion touch on it.
- Default graph options are now thread-safe.

Commit:
Improved performance, only one ObjectMapper instance per GraphResultSet.

Commit:
Improved performance and memory usage, added more checks.

- asVertex/Edge() methods moved to GraphData. this allows to transform any GraphData into a [custom]Vertex/Edge object.
- One instance of ObjectMapper is now created in a GraphResultSet, and shared to all GraphTraversalResults, and their GraphData.
|-> Except when deserialising Vertex or Edge objects (asVertex/Edge()), since the deserialisation is made out of the GraphTraversalResult. So a new ObjectMapper instance will be created on each asVertex/Edge() call. I don't see how to improve that for now.
- New comprehensive exceptions returned if the client tries to get() a non existing key or index.
- Other exceptions with clear messages.
- Other null checks.
- Optimised imports.

Commit:
Fixed minor typos, cleaned some unused fields.

Commit:
Deleted the possibility to put values in a GraphStatement constructor.

- Changed some comments.

Commit:
Disabled preparing graph statements because not ready on the Server side yet.

Commit:
Javadoc for GraphSession.

Commit:
Javadoc for Graph statements.

Commit:
Javadoc for results API.

Commit:
Recent changes to DSE Graph include : cassandra rc2, so no schema metadata conflicts anymore...

- Vertex Ids are now GraphData objects. This can chane in the future.

olim7t commented on driver-dse/src/main/java/com/datastax/driver/graph/AbstractGraphStatement.java:
```diff
@@ -0,0 +1,141 @@
+    static boolean checkStatement(AbstractGraphStatement graphStatement) {
+        return graphStatement.getGraphOptions().containsKey(GraphSession.GRAPH_SOURCE_KEY)
+            && graphStatement.getGraphOptions().containsKey(GraphSession.GRAPH_LANGUAGE_KEY)
+            && graphStatement.getGraphOptions().containsKey(GraphSession.GRAPH_SOURCE_KEY);
```
Duplicate key GRAPH_SOURCE_KEY, should probably be GRAPH_KEYPSACE_KEY (which has a typo btw).

newkek replied:
I thought I had fixed that, done.

mfiguiere commented:
I guess you guys have thought about it, but is there no way to use `*GraphStatement`s with the execute methods of the `Session`, rather than using a `GraphSession` wrapper? It's a pity to have this `GraphSession` class adapting most of the `Session` methods and adding only a few other configuration methods.

newkek commented:
We had discussed it internally indeed and it resulted that the goal of the integration of graph is to get these new Graph features without impacting the core components of the drivers. There are certainly more elegant ways to do it when modifying core components like `Session`, but this solution was chosen so we could have all the Graph components in a separate module (i.e. for Java _driver-dse_ module) and have no impact on the core of the drivers.

newkek commented:
I've just noticed that on Github the commits history is completely wrong.

Commit:
Changed behaviour of get(Object) that does not throw an exception anymore.

- GraphData behaves as follows : using get(Object) with an invalid key or index will not throw an exception, but always return a GraphData, and if the key or index used was not valid, all asX() methods will return null.
- Changed some comments and javadocs.
- Typo in GRAPH_KEYSPACE_KEY.

Commit:
Removed getResultString() method.

- In fact it has the same effect as the toString() method.

Commit:
Avoid confusion in get() methods.

- Changed the GraphTraversalResult#get() method to getRoot()
- GraphData#get(Object) to GraphData#get(String) and GraphData#get(int)

tolbertam commented on driver-dse/src/main/java/com/datastax/driver/graph/AbstractGraphStatement.java:
```diff
@@ -0,0 +1,141 @@
+        if (graphKeyspace == null || graphKeyspace.isEmpty()) {
+            throw new InvalidQueryException("You cannot set null value or empty string to the keyspace for the Graph, this field is mandatory.");
+        }
+        this.payload.put(GraphSession.GRAPH_LANGUAGE_KEY, ByteBuffer.wrap(graphKeyspace.getBytes()));
```
shouldn't this be `GraphSession.GRAPH_KEYSPACE_KEY`?

newkek replied:
Damn it I need to add unit tests for that stuff, thanks.

tolbertam commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphTraversalResult.java:
```diff
@@ -0,0 +1,100 @@
+     * @param keyOrIndex Can be either a String or a int/Integer. Must not be null, or an exception will be thrown.
+     * @return A GraphData instance containing the value for the Json entity required.
+     */
+    public GraphData get(Object keyOrIndex) {
```
what do you think about also adding a `has` method?  Since `get` delegates to `rootNode.get(x)` which throws an Exception when no key/index exists, a `has` would be a nice check-before construct.

tolbertam commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphSession.java:
```diff
@@ -0,0 +1,225 @@
+    final static Map<String, ByteBuffer> DEFAULT_GRAPH_PAYLOAD;
+
+    static {
+        GRAPH_SOURCE_KEY        = "graph-source";
```
for convenience you could just assign these when you declare them above.

Commit:
Bad copy-paste mistake in AbstractGraphStatement#setGraphKeyspace()

Commit:
Fixed DefaultEdgeDeserializer with DSE Graph ids implementations.

Commit:
Reverted the change of GraphTraversalResult#getRoot(), renamed to get().

- Minor other change.

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/BoundGraphStatement.java:
```diff
@@ -0,0 +1,116 @@
+    BoundStatement configureAndGetWrappedStatement() {
+        JsonNodeFactory factory = new JsonNodeFactory(false);
+        JsonFactory jsonFactory = new JsonFactory();
+        ObjectMapper objectMapper = new ObjectMapper();
```
ObjectMapper instances are thread-safe and should be shared; they can even be declared as singletons.

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/AbstractGraphStatement.java:
```diff
@@ -0,0 +1,141 @@
+
+    protected final Map<String, ByteBuffer> payload;
+    protected T wrappedStatement;
+    protected final GraphSession session;
```
I'm not sure it's really required to keep a reference to a session; this "ties" the statement to a session, so what happens if it gets executed by a different `Session` instance?
At first glance it seems possible to get rid of this reference:
- It's mainly used to propagate `session.getDefaultGraphOptions()` from prepared to bound statements;
- `getSession()` is never used;
- `GraphStatement` needs the session to instantiate `SimpleStatement`s – but that should change once we rebase on top of current 3.0.

newkek replied:
This was part of the changes to adapt for the 3.0 way of creating simple statements. I am correcting this today.

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/AbstractGraphStatement.java:
```diff
@@ -0,0 +1,141 @@
+
+    }
+
+    static void copyConfiguration(AbstractGraphStatement from, AbstractGraphStatement to) {
```
This method is never used.

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/BoundGraphStatement.java:
```diff
@@ -0,0 +1,116 @@
+public class BoundGraphStatement extends AbstractGraphStatement<BoundStatement> {
+
+    //  "name", "value"
+    Map<String, Object> valuesMap;
```
Should be private final.

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/BoundGraphStatement.java:
```diff
@@ -0,0 +1,116 @@
+        copyConfigFromStatement(gst);
+    }
+
+    void copyConfigFromStatement(GraphStatement gst) {
```
Several remarks:
- This method is only called by the constructor and could be inlined;
- Shouldn't the graph options be copied to `this.payload` instead, and let it be propagated to the wrapped statement when `configure()` is called?
- I would rather have the bound statement inherit its graph options from its prepared statement, whatever they are; the bound statement shouldn't directly interrogate the session.

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/BoundGraphStatement.java:
```diff
@@ -0,0 +1,116 @@
+                    throw new DriverException("Parameter : " + value + ", is not in a valid format to be sent as Gremlin parameter.");
+                }
+                objectMapper.writeTree(generator, parameter);
+                this.wrappedStatement.setString(paramNumber++, stringWriter.toString());
```
`valuesMap` is a `HashMap` and iteration order is not deterministic, so it seems a bit strange to send positional parameters this way.
Moreover, the "gremlin-driver-integration" document states:

> To pass parameters, Gremlin really just expects a Map<String,Object> where the key becomes the name of the variable in the script to bind to and the value of the Map represents the value to apply to the script for that parameter name [...] It should be noted that Gremlin doesn't support positional parameters.

So shouldn't we simply add named parameters by calling `this.wrappedStatement.setXXX(name, value)`? Why do we need to convert each variable to JSON?

newkek replied:
This was made to follow the [specs](https://github.com/riptano/bdp/blob/DSP-5175/dse-graph/docs/dev/server.adoc#dse-graph-server-extension), parameters are sent as unnamed parameters, each containing `{name:'parameter_name', value:JSON_serialised_value}`. 
I think this is related to the way the tinkerpop's gremlin server takes parameters, and since the DSE server will embed a gremlin-server, they chose this format to avoid modifying the embedded gremlin server.
And the order should not matter as the gremlin server will know how to decode the values since each name is specified in the value itself.

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/AbstractGraphStatement.java:
```diff
@@ -0,0 +1,141 @@
+     * @return This {@link com.datastax.driver.graph.AbstractGraphStatement} instance to allow chaining call.
+     */
+    public AbstractGraphStatement setOutgoingPayload(Map<String, ByteBuffer> payload) {
+        this.wrappedStatement.setOutgoingPayload(payload);
```
Shouldn't this be: `this.payload.putAll(payload)` ? It seems that payloads are not consistently kept in sync between `his.payload` and `wrappedStatement`'s payload.

newkek replied:
Yes, it should be `this.payload.putAll(payload)`, I'll correct it.
Users should only be dealing with the `AbstractGraphStatement#payload`, I was thinking about renaming it to `AbstractGraphStatement#graphOptions` to make this distinction and take a `Map<String, String>` in this method to make it simpler as well.
Indeed they are not kept synced, `AbstractGraphStatement#payload` is applied to the `wrappedStatement`'s payload only at the statement execution time.

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphStatement.java:
```diff
@@ -0,0 +1,135 @@
+/**
+ * A simple graph statement implementation.
+ */
+public class GraphStatement extends AbstractGraphStatement<SimpleStatement> {
```
I would suggest renaming to `SimpleGraphStatement` for clarity.

olim7t replied:
:+1:

newkek replied:
Ok, I thought that `SimpleGraphStatement` was a bit long for a simple statement..

tolbertam commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphStatement.java:
```diff
@@ -0,0 +1,135 @@
+     *             Please refer to Gremlin's documentation for more information.
+     * @param value Any object serializable in JSON. The type will be detected automatically at statement's execution.
+     */
+    public void set(String name, Object value) {
```
For convenience it would be nice if this returned itself, so users could chain set, i.e.:

``` java
stmt.set("i", x).set("j", y).set("k", z);
```

tolbertam replied:
Is it possible to provide a Vertex or Edge here?  Would be nice if I could use a previously returned Vertex / Edge here.

tolbertam replied:
It looks like you can express a Vertex as a String, i.e.:

```
g.V('{member_id=0, community_id=800066, ~label=comic, group_id=1}')
```

newkek replied:
`'{member_id=0, community_id=800066, ~label=comic, group_id=1}'` is a Vertex **id**, so yes we could use `set("i", vertex.getId().toString())` from a previous result.
Considering that now I made ids a GraphData in Vertex and Edge.

tolbertam replied:
I tried that and it does not work because `getId().toString()` returns the json payload (i.e.: `{member_id: "0", community_id: "800066", ~label: "comic", group_id: 
"1"}`) which gremlin doesn't like for some reason (maybe that is something that should work on the server side?)

tolbertam commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphStatement.java:
```diff
@@ -0,0 +1,135 @@
+
+    private final String query;
+    private final Map<String, Object> valuesMap;
+    private List<String> JsonParams;
```
Should rename this to 'jsonParams'

tolbertam commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphStatement.java:
```diff
@@ -0,0 +1,135 @@
+            // Avoids regenerating the Json params if the params haven't changed.
+            return;
+        }
+        this.JsonParams.clear();
```
Since jsonParams is only ever used in configureAndGetWrappedStatement, it could just be something that is returned by processValues so it doesn't need to be an instance member.

tolbertam replied:
Ah, nevermind, I just noticed that it's maintained so `processValues` doesn't create a new map every time its called.

olim7t replied:
But `processValues` is only ever called when we execute the statement, so that can be done at the last moment (could even be done in the GraphSession code). I agree that this should not be an instance member.

newkek replied:
The goal here was to avoid regenerating the `List<String> jsonParams` each time `processValues()` is called.
By not making `jsonParams` an instance member, if we call `processValues()` twice in a row and the parameters (`valuesMap.hashCode()`) haven't changed, we could avoid regenerating the `List<String> jsonParams`. If `jsonParams` becomes a returned value, we would be forced to regenerate it each time because it wouldn't have been saved anywhere.

tolbertam commented on driver-dse/src/main/java/com/datastax/driver/graph/Edge.java:
```diff
@@ -0,0 +1,137 @@
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonDeserialize(using = DefaultEdgeDeserializer.class)
+public class Edge {
```
Since edge and vertex share some characteristics (id, label, type, properties), should we have a parent class like gremlin has with [Element](http://tinkerpop.incubator.apache.org/javadocs/3.1.1-SNAPSHOT/full/org/apache/tinkerpop/gremlin/structure/Element.html)?  The other thing I wonder is are we going to end up duplicating a lot of the gremlin java API, should we just use the gremlin API proper, or is that bringing in too much (Would add gremlin as a dependency)?  (apologies if this has already been discussed).

newkek replied:
My goal initially was to avoid having a dependency on the gremlin-driver, but I think I am going to have to reconsider this..

olim7t replied:
Reusing classes from gremlin-driver definitely sounds interesting.

Their ResultSet class is probably too specific for us, but our ResultSet could be an iterable of [Result](http://tinkerpop.incubator.apache.org/javadocs/3.1.1-SNAPSHOT/full/org/apache/tinkerpop/gremlin/driver/Result.html). There's also a [GraphSONMapper](http://tinkerpop.incubator.apache.org/javadocs/3.1.1-SNAPSHOT/full/org/apache/tinkerpop/gremlin/structure/io/graphson/GraphSONMapper.html) that seems to provide out-of-the-box deserialization of classes like Vertex and Edge.

One hurdle is that gremlin-driver compiles to Java 8, so we'd have to raise the requirement on our dse module.

adutra replied:
+1 on this.

> One hurdle is that gremlin-driver compiles to Java 8, so we'd have to raise the requirement on our dse module.

We already do this for driver-extras so it's technically easy; however that might put some DSE users out of the game.

tolbertam commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphSession.java:
```diff
@@ -0,0 +1,225 @@
+     * @return A {@link GraphResultSet} object if the execution of the query was successful. This method can throw an exception
+     * in case the execution did non complete successfully.
+     */
+    public GraphResultSet execute(AbstractGraphStatement statement) {
```
I think we need an interface above `AbstractGraphStatement` or we need to make it public since this is a public method, there could be weird classloading issues (BootstrapMethodError) if someone tries to use this as a method reference in java 8.

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphData.java:
```diff
@@ -0,0 +1,215 @@
+        return new GraphData(keyOrIndex, jsN, this.objectMapper);
+    }
+
+    public GraphData get(String key){
```
One should clarify that get(String) applies to JSON objects and get(int) to JSON arrays - calling the wrong one returns null (took me a while to figure out).

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphData.java:
```diff
@@ -0,0 +1,215 @@
+     *
+     * @return The enclosed JSON object.
+     */
+    public JsonNode getJsonObject() {
```
If we are exposing Jackson API through our own API, I don't really see the point of wrapping a Jackson JsonNode object in a GraphData object.

newkek replied:
I was thinking of leaving the possibility for more advanced users to do more advanced stuff we would not want to provide in the drivers because it would not fit most use cases, I don't have a clear example right now but I guess it could happen

adutra commented on driver-dse/src/main/java/com/datastax/driver/graph/GraphData.java:
```diff
@@ -0,0 +1,215 @@
+
+    private ObjectMapper objectMapper;
+
+    private Object key;
```
This seems unused.

Commit:
Removed all new instanciations of a ObjectMapper, kept only one in GraphSession.

Commit:
Rebased on current 3.0.

- SimpleStatement constructor public.
- No Session instance in the AbstractGraphStatement and its child classes.
- Prepared and BoundGraphStatements structure improved and tested (unit tests to insure graph options are propagated and parameter values are bound correctly)
- Minor other improvements.

Commit:
Optimized imports and reformated on default java scheme.

Commit:
Various minor changes from reviews comments.

newkek commented:
Latest update on this is : 
- Removed all new instantiations of a ObjectMapper, kept only one in GraphSession 9a50dd1
- Rebased on current 3.0. 3947753 : 
- GraphStatement constructor public.
- No Session instance in the AbstractGraphStatement and its child classes.
- Prepared and BoundGraphStatements structure improved and tested (unit tests to insure graph options are propagated and parameter values are bound correctly)
- Optimised imports and reformatted on default java scheme. e244966
- Various minor changes from reviews' comments. f8aea24

adutra commented:
Closing because this code is being moved out of the java-driver repo.
